### PR TITLE
Writing file_paths to table and improved FilePaths, FileRevs, Hunks and HunkBlame

### LINKED ISSUE #66: [Bug] Hunks/FilePaths memory issues
This is somewhat related to issue 37, however the problem is elsewhere.

Hunks calls fp.update_all(repo_id) in (line 277 develop branch), which eventually causes a crash after running out of memory. update_all in turn calls update_for_revision for all commit_ids, which ultimately fails at line 128 (which as far as I gather is filling the cached adjacency matrix)

Tested with amarok git repository, on a 4GB Core 2 Duo, it fails after about 7500 revisions.

I have no idea what the solution to this could be, but I guess splitting into chunks could be one way to go about it...

### PR DESCRIPTION
**Pull request by apepper**
This should fix #66.

### CODE REVIEW

**Comment by cflewis** [comment]
Please do not merge without approval from linzhp and Lewisham.

## 

Chris Lewis (on cell phone, apologies for typos)
On May 13, 2011 4:10 AM, "apepper" <
reply@reply.github.com>
wrote:

File: docs/cvsanaly.mdown [hunk 1]
```diff
@@ -234,6 +234,15 @@ The `file_copies` table is used to store additional information about actions th
 * `new_file_name`: contains the new name of the file for rename actions or `NULL` for other actions. 
 * `action_id`: the identifier of the action. This is a foreign key that references the `id` field of the `actions` table.
 
+#### file_paths table
+
+The `file_paths` table is used to eighter look up the `file_id` for a given `file_path` and `commit_id` or to look up the current `file_path` for a given `file_id` and `commit_id`. This is an alternative table to files, `file_links` and `actions`.
```
Comment (cflewis) [inline docs/cvsanaly.mdown]: Typo: eighter -> either.

File: pycvsanaly2/DBContentHandler.py [hunk 2]
```diff
@@ -214,6 +214,20 @@ class DBContentHandler(ContentHandler):
                               dbfilecopy.new_file_name, 
                               dbfilecopy.action_id))
 
+    def __add_file_path(self, commit_id, file_id, path):
+        try:
+            file_path = path.split("://", 1)[1]
+        except IndexError:
+            file_path = path
+
+        dbfilepath = DBFilePath(None, commit_id, file_id, file_path)
```
Comment (cflewis) [inline pycvsanaly2/DBContentHandler.py]: Please rename `dbfilepath` to `db_file_path`. This looks more "Pythonic". Please change this throughout the file.

File: pycvsanaly2/extensions/FilePaths.py [hunk 3]
```diff
@@ -144,10 +144,36 @@ class FilePaths:
...
 
+    def get_path2(self, file_id, commit_id, repo_id):
+        if config.debug:
+            profiler_start("Getting full file path for file_id %d and commit_id %d", (file_id, commit_id))
+
+        db = self.__dict__['db']
+        cnn = db.connect()
+
+        cursor = cnn.cursor()
+        query = "SELECT file_path from file_paths " + \
```
Comment (cflewis) [inline pycvsanaly2/extensions/FilePaths.py]: Please modify the query to be a fully quoted string, using three double-quotes. eg.

```
"""SELECT file_path from file_paths"""
```

Don't worry about whitespace, SQL is whitespace-insensitive. This makes it easier to edit the query string later.

File: pycvsanaly2/extensions/FilePaths.py [hunk 4]
```diff
@@ -144,10 +144,36 @@ class FilePaths:
...
+
+        cursor.close()
+        cnn.close()
+
+        if config.debug:
+            profiler_stop("Getting full file path for file_id %d and commit_id %d", (file_id, commit_id), True)
+            printdbg("get_path2: Path for file_id %d at commit_id %d: %s", (file_id, commit_id, file_path))
+
+        for file_path in all_file_paths:
+            return file_path
```
Comment (cflewis) [inline pycvsanaly2/extensions/FilePaths.py]: What's being returned here?

`all_file_paths` is a list, which I am guessing you are assuming contains just one item. If so, put in a comment to that effect and just `return all_file_paths[0]`. I think the `for` loop here will just stop after the first iteration. If you are expecting to return a list, then do `return all_file_paths`.

I think this method should only return one string, and if the `all_file_paths` list has more than one element in it, an exception should be thrown.

File: pycvsanaly2/extensions/FilePaths.py [hunk 5]
```diff
@@ -144,10 +144,36 @@ class FilePaths:
             #use get instead of index to avoid key error
             id = adj.adj.get(id) 
 
-            # profiler_stop("Building path for file %d", (file_id,), True)
+        profiler_stop("Building path for file %d", (file_id,), True)
 
         return "/" + "/".join(tokens)
 
+    def get_path2(self, file_id, commit_id, repo_id):
```
Comment (cflewis) [inline pycvsanaly2/extensions/FilePaths.py]: Is there a better name for this than `get_path2`? Maybe `get_path_from_database`?

Also, please add a [docstring](http://www.python.org/dev/peps/pep-0257/) to this method.

@linzhp and I have been bad about doing docstrings, but I'm going to begin enforcing docstrings on new code going in, as its becoming increasingly difficult to know what is happening :)

File: pycvsanaly2/extensions/FilePaths.py [hunk 6]
```diff
@@ -180,6 +206,32 @@ class FilePaths:
         except KeyError:
             return None
 
+    def get_file_id(self, file_path, commit_id):
+        if config.debug:
```
Comment (cflewis) [inline pycvsanaly2/extensions/FilePaths.py:216]: Docstring, please :)

File: pycvsanaly2/DBContentHandler.py [hunk 7]
```diff
@@ -214,6 +214,20 @@ class DBContentHandler(ContentHandler):
                               dbfilecopy.new_file_name, 
                               dbfilecopy.action_id))
 
+    def __add_file_path(self, commit_id, file_id, path):
+        try:
```
Comment (cflewis) [inline pycvsanaly2/DBContentHandler.py:220]: Please add a docstring (explanation in comment elsewhere in diff).

File: pycvsanaly2/extensions/FileRevs.py [hunk 8]
```diff
@@ -105,7 +104,7 @@ class FileRevs:
             rev = revision
 
         try:
-            relative_path = self.fp.get_path(file_id, commit_id, 
+            relative_path = self.fp.get_path2(file_id, commit_id,
```
Comment (cflewis) [inline pycvsanaly2/extensions/FileRevs.py]: Will require renaming if `get_path2` is renamed.

File: pycvsanaly2/extensions/HunkBlame.py [hunk 9]
```diff
@@ -39,10 +39,6 @@ class HunkBlameJob(Job):
```
Comment (cflewis) [inline pycvsanaly2/extensions/HunkBlame.py:38]: Will require review from @linzhp.

File: pycvsanaly2/extensions/Hunks.py [hunk 10]
```diff
@@ -270,55 +271,20 @@ class Hunks(Extension):
...
-                                    file_id = possible_file[0]
-                                    break
-                                    break
-                           
-                            if possible_file[1] == hunk_file_name:
-                                file_id = possible_file[0]
-                                break
-                                break
-    
+                    file_id = fp.get_file_id(hunk_file_name, commit_id)
```
Comment (cflewis) [inline pycvsanaly2/extensions/Hunks.py:286]: Gosh, seeing this makes me so happy :)

**Comment by cflewis** [comment]
Hi @apepper,
Thanks for this commit, it looks awesome. I've been nitpicky in the diffs. I've generally been harder on you than we have been on ourselves (such as docstrings), but I am trying to be harder on ourselves too after seeing what is working in industry on my internship. I know this is a bit hypocritical, for which I apologize!

Please don't take my comments that your code was bad; it was really good (especially for a Python novice) and I am looking forward to trying this out!

Once commits that address the comments are in this pull request, I will give approval to merge. Because this is our first real, feature-adding, super-exciting pull request, I will also be waiting for approval from @linzhp too.

File: pycvsanaly2/extensions/HunkBlame.py (hunk 9 above)
Comment (linzhp) [inline pycvsanaly2/extensions/HunkBlame.py:38]: I reviewed this change this morning, and actually this fixed an error that broke cvsanaly I ran on tortoise over last night. I would have merged this change, but after I search the internet on how to do that, I found someone said that was a bad practice to skip other commits and merge a commit alone.

Comment (cflewis) [inline pycvsanaly2/extensions/HunkBlame.py:38]: Yeah, cherry-picking commits is just going to make it harder to do the merge later, you made the right call. You can, of course, bring the merge into a fork in your own repo.

**Comment by apepper** [comment]
I changed the requested things. Please review and pull if you aprove it.

Greetings from Berlin
Alex

**Comment by cflewis** [comment]
Looks good to me! It's got my approval.

Will wait for approval from @linzhp, then he or I can do the merge and we'll schedule an immediate test run on our [Jenkins install](http://eisbox.soe.ucsc.edu/jenkins/job/cvsanaly/) to make sure it works as expected (and see how much quicker it can run!).

**Comment by cflewis** [comment]
Debug level run is [going now](http://eisbox.soe.ucsc.edu/jenkins/job/cvsanaly/137/console). It looks like its working. Awesome!

**Pull request merged**